### ISSUE TITLE
Possibility to prioritize source of popup window

### ISSUE THREAD

**Issue by sassanh** [roxma/nvim-completion-manager#116]
Currently if in a python buffer I type:
```
import requests
requests.pos
```
and select "post" from autocompletion list popup window shows:
```
snippet@0
```
If I disable `ultisnips` and restart vim I get
```
post(url, data=None, json=None, **kwargs)
    Sends a POST request.

    :param url: URL for the new :class:`Request` object.
    :param data: (optional) Dictionary (will be form-encoded), bytes, or file-like object to send in the body of the :class:`Request`.
    :param json: (optional) json data to send in the body of the :class:`Request`.
    :param \*\*kwargs: Optional arguments that ``request`` takes.
    :return: :class:`Response <Response>` object
    :rtype: requests.Response
```

Is it possible to achieve this without disabling ultisnips? some way to prioritize python help over ultisnips for popup window.

**Comment by roxma** [comment]
The `snippet@0` is a workaround for snippet expansion, it is appended to the preview window, it does not control the display of the docstring `Sends a POST request.` if it exists.

It seems this is a behavior of the jedi library.

Here's my test on jedi-vim's omnifunc with `set completeopt=menu,menuone,noinsert,noselect,previe`:

Test 1:

```python
import requests
requests.<c-x><c-o>
```

Then select the `post` item, the preview window does not appear

Test 2:

```python
import requests
requests.pos<c-x><c-o>
```

Then select the `post` item, the preview window appears

**Comment by sassanh** [comment]
Well I think that's another issue or maybe I didn't get the relation. To clarify: my problem is when I have ultisnips disabled it shows doc string but when ultisnips is enabled instead of docstring it shows `snippet@0`. I want to somehow have docstring even when ultisnips is enabled.

**Comment by sassanh** [comment]
I confirm even when ultisnips is disabled it doesn't show docstring when I type `request.p` and then select `post`. But with ultisnips disabled it does show docstring with `request.pos` and then selecting `post` and when ultisnips is enabled it shows `snippet@0` even when I type `request.pos` and then select `post`.

**Comment by roxma** [comment]
> Well I think that's another issue or maybe I didn't get the relation

When the docstring return by jedi is empty, but there's a snippet for the item, it only shows the snippet.

>  I want to somehow have docstring even when ultisnips is enabled.

Ultisnips should not affect the display of docstring if it's not empty. And I cannot reproduce the issue you described.

![peek 2017-08-18 10-13](https://user-images.githubusercontent.com/4538941/29441712-c22d125a-83ff-11e7-8883-9b7b5ca39f23.gif)

**Comment by roxma** [comment]
> and when ultisnips is enabled it shows snippet@0 even when I type request.pos and then select post.

NCM cache the result when you typed `.`, in this case, you need to force refresh with the  `<Plug>(cm_force_refresh)` key when the cursor is at `pos|`.

**Comment by roxma** [comment]
This gif shows my point.

![peek 2017-08-18 10-13](https://user-images.githubusercontent.com/4538941/29441966-6ec04aea-8401-11e7-80b4-371ecd1edcfa.gif)

**Comment by sassanh** [comment]
Yeah seems like it's related to cache. Is it possible to disable cache altogether?

**Comment by roxma** [comment]
>  Is it possible to disable cache altogether?

Not possible with current version. The cache control is hard-coded by ncm's jedi source.

**Comment by sassanh** [comment]
OK, thanks for info.

**Comment by roxma** [comment]
I decided to disable the caching of jedi source.

You could test this PR #128

**Comment by roxma** [comment]
This is how it looks like. I believe it should still be considered an upstream bug.

![116](https://user-images.githubusercontent.com/4538941/29864340-eeb5c584-8da4-11e7-8a04-866389d4fdf9.gif)

**Comment by sassanh** [comment]
Unfortunately it doesn't work for me:
https://asciinema.org/a/5XU59J75bDiC3pFgpwj6tu17S

**Comment by roxma** [comment]
It seems the no-cache version slows things down.

And unfortunately this will make fuzzy matching impossible, since jedi doesn't have builtin fuzzy matching: https://github.com/davidhalter/jedi/issues/827

**Comment by roxma** [comment]
Maybe I should revert the change.

**Comment by sassanh** [comment]
Yeah seems like we should wait for jedi.

**Comment by roxma** [comment]
I've pushed another update. Please test it in case the no-cache is not the cause.

**Comment by sassanh** [comment]
It's working for me now. Thanks.
